Display emoji descriptions in popups

eranl opened the pull request:
This is a draft for discussion.

Using descriptions downloaded from CLDR, an asset file per supported locale is created.
The script that creates these files depends on #1515 for the lists of supported emojis.
These files are bundled with the app. Should they be made into separate download(s)?
Using modified popups, these descriptions are then shown for emoji keys.
Some of the adaptations I made to popups seem a little hacky to me, so I'd appreciate advice for improving this.
This feature should probably be controlled by a setting.
emojis with flavors don't show descriptions, since the popups are taken. I'd appreciate ideas for fixing this.
Saving and restoring recent emojis loses the descriptions. To be fixed.
A `?` artifact shows in the popups, which is weird. To be fixed.

Fixes #1532.

Helium314 commented:
> The descriptions files are bundled with the app. Should they be made into separate download(s)?

Hm. Yes and no?
It's a _lot_ of additional data (11.5 MB diff, 2 MB larger APK), which is too much for a feature no one appeared to be interested in until #1532.
But the descriptions probably don't work with a dicitonary (though I didn't try), when adding them to an emoji dictionary would be the most viable alternative.
But if it's not in dictionaries, asking users to download yet another thing seems a little excessive... still I thing bundling the descriptions with the app is too much.

> Some of the adaptations I made to popups seem a little hacky to me, so I'd appreciate advice for improving this.

Popups are... weird. I understand they were made the way they are for performance reasons, but dealing with them can be awful.
Some description text is even too long for the popups.
It might be better to use some other view than a popup, like a TextView with some sort of border.

> Descriptions are shown in the current keyboard language. Should we use the main system language instead? I can see reasons to do either.

I tend towards the keybord language, but no strong opinion here.

> Emojis with flavors don't show descriptions, since the popups are taken. I'd appreciate ideas for fixing this.

Currently my only idea would be displaying the text in a separate line. But that would look rather awful, and you'd run into the "all popups are same size" problem.

eranl commented:
Thanks for the feedback.

> a feature no one appeared to be interested in until #1532.

I think most people don't know it's possible, because they haven't seen it anywhere. I certainly didn't know it was possible until I found CLDR.

> asking users to download yet another thing seems a little excessive... still I thing bundling the descriptions with the app is too much.

So is this feature feasible? Maybe bundle descriptions with dictionaries?

Helium314 commented:
> So is this feature feasible? Maybe bundle descriptions with dictionaries?

I don't know how they could be resonably bundled with the dictionaries without throwing overboard the idea of making dicitonaries for all AOSP keyboards.
Maybe there is a way to have them in the dictionaries?

> like a TextView with some sort of border.

Actually if such a TextView is positioned properly, it could be made to appear above the popups, so you simply display both.

eranl commented:
> Maybe there is a way to have them in the dictionaries?

Based on [this](https://codeberg.org/Helium314/aosp-dictionaries/src/branch/main/README.md#wordlist-information), I don't see how. Do you have an idea for doing that?

> Actually if such a TextView is positioned properly, it could be made to appear above the popups, so you simply display both.

So either one or both of these would show up on long tap? Should I make any changes to the hint?

Helium314 commented:
> Do you have an idea for doing that?

Currently my only idea would be trying to mis-use the shortcut, like `word=👋🏻` and `shortcut=this is a hand`. But I didn't try, maybe it doesn't even compile, maybe it causes issues somewhere else.

> So either one or both of these would show up on long tap?

I'd say both, so on top you have the view with the description, and below the usual popup keyboard.

> Should I make any changes to the hint?

If this is added via setting, I think the users should be aware and a hint might not be necessary.

eranl commented:
> I'd say both, so on top you have the view with the description, and below the usual popup keyboard.

Either one of these may not be available for any given emoji.

> If this is added via setting, I think the users should be aware and a hint might not be necessary.

The thing is not all emojis have descriptions available.

Helium314 commented:
> The thing is not all emojis have descriptions available.

Is this still valid with the CLDR descriptions? I'm considering packaging at least the English dictionary with Heliboard, which could be used as fallback. (or do users actually expect descriptions for the "emoticons" section?)

> Either one of these may not be available for any given emoji.

When there are no popups the description text could just be placed lower?
And both not existing just shows nothing, right?

eranl commented:
> > The thing is not all emojis have descriptions available.
> 
> Is this still valid with the CLDR descriptions?

Yes, not all emojis have CLDR entries, and not all entries have tts .

> When there are no popups the description text could just be placed lower? And both not existing just shows nothing, right?

Done.

eranl commented:
Questions about my latest commit:
1. Any idea about the build failure? I don't get it.
2. I made `AndroidSpellCheckerService.createKeyboardForLocale` `public static` in order to reuse it. Is there a utility class I can move it to?
3. Do I need to use exclusive numbered dictionary sessions? I just used `0` with no guard around it for now.
4. With description frequencies set to `255`, checking suggestion results for `KIND_WHITELIST` or `KIND_SHORTCUT` and `mScore > 0` seems to catch all true descriptions, and only a few non-descriptions. Anything I can do to filter the latter out?

Helium314 commented:
1. my fault, I had put the wrong hash in `gradle-wrapper.properties`
2. why do you create another keyboard?
3. I never worked with sessions for the dictionaries. I guess this might be necessary when there are parallel calls to `getSuggestions` on the same dictionary? For testing you could try triggering issues by doing multiple calls in parallel on the same sessionId. Or you could just leave it at 0 and we deal with it in case we encounter problems.
4. which dictionaries are you querying?
5. hmm... when the setting is opt-in we could assume the user wants to check descriptions and not inserting is definitely perferable. But as it's now I'm not sure.

eranl commented:
> why do you create another keyboard?

What's the alternative?

> I guess this might be necessary when there are parallel calls to `getSuggestions` on the same dictionary? For testing you could try triggering issues by doing multiple calls in parallel on the same sessionId. Or you could just leave it at 0 and we deal with it in case we encounter problems.

I did notice some weird cases with rapid successive lookups, where one lookup seemed to 'bleed' into another. Where do I get info about how many sessions I should have?

> which dictionaries are you querying?

I guess all, as I didn't see a way to limit it. But the matches seem to all come from the emoji dictionary - they seem to be manufactured from other entries.

Helium314 commented:
> What's the alternative?

I haven't properly looked through the code yet, but my idea way getting some kind of offset from the popup keyboard (maybe whether popups are displayed or not is already enough). The offset could be added to the view.
I'll need to find some time for review before I can give a useful comment on this.

> I did notice some weird cases with rapid successive lookups, where one lookup seemed to 'bleed' into another. Where do I get info about how many sessions I should have?

Sorry, I don't have any experience on this. You will need to test and find something that works.

> I guess all, as I didn't see a way to limit it

It should really only be the emoji dictionary. I think at some other place I told you of my (halfway done) plan to have some kind of DictionaryFacilitator (or at least something similar) that works with a single dictionary. This would be ideal for what you need.

eranl commented:
> my idea way getting some kind of offset from the popup keyboard (maybe whether popups are displayed or not is already enough). The offset could be added to the view.

What can I use for the `keyboard` parameter [here](https://github.com/Helium314/HeliBoard/blob/18549151b38f5b398f60d30c3cee690038080e7c/app/src/main/java/helium314/keyboard/latin/DictionaryFacilitator.java#L130-L131)?

> my (halfway done) plan to have some kind of DictionaryFacilitator (or at least something similar) that works with a single dictionary.

Please let me know when it's ready, or if I can help with that.

eranl commented:
Is it a bug that adding/removing/replacing a dictionary requires an app restart?

Helium314 commented:
> Is it a bug that adding/removing/replacing a dictionary requires an app restart?

That would be a bug.

> Please let me know when it's ready, or if I can help with that.

It's added in e034065236d54d2b3108b4dc48d06816f6c948cb (original intention was the thing that will be using the `suggestionLogger`)

> What can I use for the keyboard parameter [here](https://github.com/Helium314/HeliBoard/blob/18549151b38f5b398f60d30c3cee690038080e7c/app/src/main/java/helium314/keyboard/latin/DictionaryFacilitator.java#L130-L131)?

https://github.com/Helium314/HeliBoard/blob/main/app/src/main/java/helium314/keyboard/latin/SingleDictionaryFacilitator.kt#L20

eranl commented:
> https://github.com/Helium314/HeliBoard/blob/main/app/src/main/java/helium314/keyboard/latin/SingleDictionaryFacilitator.kt#L20

Thanks. How do I get a `Dictionary` to pass to it?

eranl commented:
> after you long-press an emoji and a description (only) pops up, the emoji does _not_ get inserted.

I fixed this inconsistency - now the emoji always gets inserted after a long press.

> I did notice some weird cases with rapid successive lookups, where one lookup seemed to 'bleed' into another.

This is still a (rare) problem. An example: if I long-press *judge*, then *man judge*, the popup will show *judge*. If I long-press a second time, the popup is correct, but then it will bleed to the next popup. In other cases, no description is shown on the first long press after seeing a description of a different emoji.

Before switching to `SingleDictionaryFacilitator`, I experimented with random session ids, and that fixed the bleeding. But I don't know the semantics of sessions.

> You can use `DictionaryInfoUtils.getCachedDictsForLocale` and then create a `ReadOnlyBinaryDictionary` from the file. See `DictionaryFactory.checkAndAddDictionaryToListNewType` for an example.

Thanks. Done. I refactored `DictionaryFactory.checkAndAddDictionaryToListNewType` in order to reuse the logic. It's less efficient now in the case of multiple dictionaries for a locale-type combination, but that's rare, right?

> If the dictionary is in assets, you need to extract it first using `checkAndAddDictionaryToListNewType`, then it's found by `getCachedDictsForLocale`.

This is done on startup, right?

Helium314 commented:
> Thanks. Done. I refactored DictionaryFactory.checkAndAddDictionaryToListNewType in order to reuse the logic. It's less efficient now in the case of multiple dictionaries for a locale-type combination, but that's rare, right?

It's rare, and I heavily doubt the change can a relevant impact on performance at all.

> This is done on startup, right?

Dictionaries are extracted only when they are needed, i.e. we have a dictionary in assets with a type that is not available in the cached dictionaries (`DictionaryFactory` line 37).

> This is still a (rare) problem. An example: if I long-press judge, then man judge, the popup will show judge. If I long-press a second time, the popup is correct, but then it will bleed to the next popup. In other cases, no description is shown on the first long press after seeing a description of a different emoji.

I'll have a look at this, but I also have no experience with session ids and will need to investigate.

eranl commented:
> Dictionaries are extracted only when they are needed, i.e. we have a dictionary in assets with a type that is not available in the cached dictionaries (`DictionaryFactory` line 37).

Am I guarantied that this has been done by the time the emoji keyboard is shown, or do I have to trigger it? How?

Helium314 commented:
> Am I guarantied that this has been done by the time the emoji keyboard is shown, or do I have to trigger it? How?

Well, the file is extracted after `DictionaryInfoUtils.extractAssetsDictionary` has finished. So if you call it before showing the emoji keyboard then yes.  But you will need to call it yourself when you want to have a dicitonary file that's only on assets.
Note that this call does now return null on failure to extract, related to #801 (but only should happen before device is unlocked).

eranl commented:
Not sure I follow. You mean that if we add an emoji dictionary to assets, and the user doesn't add it to their list of dictionaries, I'll have to do that?

Helium314 commented:
> You mean that if we add an emoji dictionary to assets, and the user doesn't add it to their list of dictionaries

At the moment the user doesn't have a choice, all dictionaries reasonably matching the locale will be used and thus extracted.
But this should be optional when we include an emoji dictionary. So you should consider the case the dictionary exists in assets only and needs to be extracted before it can be used.

Helium314 commented:
I'd say the descriptions look mostly fine, but the text is clearly too large. Text size should be closer to what we see in clipboard view. It's not just style, but also because long descriptions are cut off, and popup positon sometimes is shifted when the description is long.

Helium314 commented:
> This is still a (rare) problem. An example: if I long-press judge, then man judge, the popup will show judge. If I long-press a second time, the popup is correct, but then it will bleed to the next popup. In other cases, no description is shown on the first long press after seeing a description of a different emoji.
>
> Before switching to SingleDictionaryFacilitator, I experimented with random session ids, and that fixed the bleeding. But I don't know the semantics of sessions.

I tried copying the behavior of `AndroidSpellCheckerService` with the session id pool that essentially results in alternating sessions 0 and 1. This seems to fix the issue.
Now I really wonder whether this should be done everywhere. Currently the session id is always 0, except when used from the spell checker. The name `MAX_NUM_OF_THREADS_READ_DICTIONARY` in `AndroidSpellCheckerService` implies it's connected to threads, but actually we only use a single thread for getting emoji desriptions.

Helium314 commented:
Weirdly enough this suggestion bleeding happens for some emojis (e.g. man / woman judge, man / woman health worker), but not for others (student, teacher, cook, ...).

Anyway, I checked normal (typing / gesturing) suggestions with different session ids, and here it looks like the session id doesn't matter, which is a little unexpected.

But I found that when don't always use the same coordinate for `ComposedData.createForWord` then the bleeding stops. Currently `NOT_A_COORDINATE` is used because it appears to make sense. Replacing just one of the coordinates with `Random.nextInt(2)` is fine (might still bleed rarely I assume).
So looks like the issue is related to different words using the same coordinates and we're stumbling over some sort of cache. Changing `ComposedData.createForWord` to use random coordinates (in a very limited range) seems to be the best fix.

Now I wonder whether this could cause issues in other places... but at least in most cases `NOT_A_COORDINATE` is used, it doesn't reach the dicitonary facilitator.

eranl commented:
> the text is clearly too large. Text size should be closer to what we see in clipboard view.

Is it better now?

eranl commented:
> Maybe the popup should be right-justified when it's close to the right edge (when the default selected popup key is right of center)?

Done.

Also:
1. Noticed that the emoji popup has no vertical gap above the key like other popups, which makes it hard to see, so added some.
2. Realized that the description popup wasn't centered horizontally over the key like other popups. Fixed.

Helium314 commented:
Are you done with changes for now? I only got about halfway through with the review on Tuesday, but then didn't have enough time to continue... Anyway, I'd just like to make sure the code doesn't change during review.

Helium314 commented:
Great, I hope to be done this weekend.
Merge likely will need to wait a little, as I plan to release another beta and then full version a week later. This is mostly to find issues with the recent internal changes, but also to make the broken backups finally work again.

Helium314 commented on app/src/main/res/layout/popup_keys_keyboard.xml:
```diff
@@ -10,6 +10,14 @@
+        android:id="@+id/description_view"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:paddingHorizontal="5dp"
```
I think the padding could be a bit larger. What do you think about 8 dp horizontal and 4 dp bottom padding? (from testing it looks like we don't need top padding)

eranl replied:
Done.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/PopupTextView.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * A view that displays popup text.
+ */
+public class PopupTextView extends TextView implements PopupKeysPanel {
```
Maybe rename it to `PopupDescriptionView`? It more specific describes what it's actually used for, and I don't think we will use it for another purpose (or do you have some plans?).

eranl replied:
There's nothing about this class that's specific to description, so I thought the name should reflect just what it is, and not what it's currently used for.

Helium314 replied:
I disagree since you add the whole thing for one specific purpose and seem to have no intention to use it for anything else. If it had more uses than "hey we could display some text on longpressing any key, but why would we?" I''d agree.
But whatever

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPageKeyboardView.java:
```diff
@@ -209,6 +217,16 @@ public boolean isShowingPopupKeysPanel() {
+        var layoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,
+                                                   ViewGroup.LayoutParams.WRAP_CONTENT);
+        layoutParams.gravity = layoutGravity != Gravity.NO_GRAVITY? layoutGravity : mDefaultLayoutGravity;
+        mPopupKeysKeyboardContainer.setLayoutParams(layoutParams);
+        mDescriptionView.setLayoutParams(layoutParams);
+        mPopupKeysKeyboardView.setLayoutParams(layoutParams);
```
Is it possible to avoid overwriting the layouts params set in the xml? It's not a horrible thing, but just might be confusing because the values in xml now aren't used in the emoji keyboard.

eranl replied:
I agree, but I couldn't find a way to set just gravity.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPageKeyboardView.java:
```diff
@@ -74,12 +81,16 @@ public void onReleaseKey(final Key key) {}
     // More keys panel (used by popup keys keyboard view)
     // TODO: Consider extending to support multiple popup keys panels
     private PopupKeysPanel mPopupKeysPanel;
+    private SingleDictionaryFacilitator mDictionaryFacilitator;
```
I don't like having the dictionary facilitator everywhere... I'd prefer keeping it in EmojiPalettesView and having the keyboard veiew ask for descriptions using the `OnKeyEventListener`.

eranl replied:
Done. Should we rename the interface to something like `EmojiViewCallback`?

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPalettesView.java:
```diff
@@ -184,6 +188,7 @@ private void updateState(@NonNull RecyclerView recyclerView, PagerViewHolder vie
 
     private final EmojiCategory mEmojiCategory;
     private ViewPager2 mPager;
+    private SingleDictionaryFacilitator mDictionaryFacilitator;
```
Maybe the facilitator could be stored somewhere else, or made static? Currently when the view is destroyed (e.g. on theme change or orientation change) the facilitator is not closed, and on these events there is no need for a reload anyway.

eranl replied:
Does `stopEmojiPalettes` get called in those cases?

Helium314 replied:
No, and I think you can't ensure it's called.

eranl replied:
Done.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/LatinIME.java:
```diff
@@ -1760,8 +1760,7 @@ private void loadKeyboard() {
             // Reload keyboard because the current language has been changed.
-            mKeyboardSwitcher.loadKeyboard(getCurrentInputEditorInfo(), mSettings.getCurrent(),
-                    getCurrentAutoCapsState(), getCurrentRecapitalizeState());
+            mKeyboardSwitcher.reloadKeyboard();
```
The difference here is only `mEmojiPalettesView.clearKeyboardCache()`, right? This seems a bit overkill, as we just want to reload the facilitator in case the language has changed (which we can assume is true when we come here).
Maybe `clearKeyboardCache` and something like `resetFacilitator` should be separate?

eranl replied:
Reverted, as this change is now unnecessary. 
Also refactored the multiple almost-identical calls to `loadKeyboard`. Let me know if that's ok.

eranl commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPalettesAdapter.java:
```diff
@@ -50,7 +55,6 @@ public void onBindViewHolder(@NonNull EmojiPalettesAdapter.ViewHolder holder, in
         final Keyboard keyboard =
                 mEmojiCategory.getKeyboardFromAdapterPosition(mCategoryId, position);
         holder.getKeyboardView().setKeyboard(keyboard);
-        holder.getKeyboardView().setOnKeyEventListener(mListener);
```
Because it's the correct place - this line is unrelated to binding to a specific position/category.
This change has nothing to do with this PR - it was just incidental cleanup.
Should I revert?

Helium314 replied:
No need to rever, I was just interested in whether it was necessary for some of the other changes.

eranl commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPalettesView.java:
```diff
@@ -414,6 +418,22 @@ public void clearKeyboardCache() {
+        if (Settings.getValues().mShowEmojiDescriptions) {
+            var locale = RichInputMethodManager.getInstance().getCurrentSubtype().getLocale();
+            var dictFile = DictionaryInfoUtils.getCachedDictForLocaleAndType(locale, "emoji", getContext());
+            mDictionaryFacilitator = dictFile != null?
```
I don't think falling back to English is a good idea if the user doesn't have English as one of their languages.
Falling back to another language is problematic also because using those descriptions in emoji search won't work, unless we fall back there too, which would be strange...
The user can always switch languages, or install another dictionary, so I think we should keep it simple and consistent.

eranl commented on app/src/main/java/helium314/keyboard/keyboard/PopupKeysKeyboardView.java:
```diff
@@ -157,6 +157,8 @@ private void showPopupKeysPanelInternal(final View parentView, final Controller
         mOriginX = x + container.getPaddingLeft();
         mOriginY = y + container.getPaddingTop();
+        var center = panelX + getMeasuredWidth() / 2;
+        controller.setLayoutGravity(center < pointX? Gravity.RIGHT : center > pointX? Gravity.LEFT : Gravity.NO_GRAVITY);
```
> Is it possible (and reasonably little work!) to change the gravity only if the text is wirder than the popup panel?

Done (in `EmojiPageKeyboardView`).

> Oh, and could you add a comment on why this line is added?

Done.

eranl commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPageKeyboardView.java:
```diff
@@ -74,12 +81,16 @@ public void onReleaseKey(final Key key) {}
     // TODO: Consider extending to support multiple popup keys panels
     private PopupKeysPanel mPopupKeysPanel;
+    private SingleDictionaryFacilitator mDictionaryFacilitator;
+    private final int mDefaultLayoutGravity;
```
Changed default gravity to center. Does it look ok to you? I thought the two views should be justified, but this is definitely simpler.

eranl commented on app/src/main/java/helium314/keyboard/settings/screens/PreferencesScreen.kt:
```diff
@@ -111,6 +112,14 @@ fun createPreferencesSettings(context: Context) = listOf(
+        R.string.show_emoji_descriptions_summary
+    ) {
+        SwitchPreference(it, Defaults.PREF_SHOW_EMOJI_DESCRIPTIONS) {
+            KeyboardSwitcher.getInstance().setThemeNeedsReload()
```
Done.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPalettesView.java:
```diff
@@ -414,6 +418,22 @@ public void clearKeyboardCache() {
         }
 
         mEmojiCategory.clearKeyboardCache();
+        initDictionaryFacilitator();
```
Reload is now not happening when emoji keyboard isn't used at all, but it's happening every time you enter emoji view.
One idea is on `initDictionaryFacilitator` check whether the current facilitator is for the current locale. If it is, then no need to reload. And then `stopEmojiPalettes` of course shouldn't close the facilitator.

eranl replied:
> `stopEmojiPalettes` of course shouldn't close the facilitator.

Should I move the closing back to `clearKeyboardCache` then?

eranl replied:
Done.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/emoji/EmojiPalettesView.java:
```diff
@@ -302,6 +305,20 @@ public void onReleaseKey(final Key key) {
+        }
+
+        var wordProperty = mDictionaryFacilitator.getWordProperty(emoji);
+        if (! wordProperty.mHasShortcuts) {
```
`wordProperty` can be null. Android Studio should actually warn about this.

Helium314 commented:
2 minor things left:
* `DictionaryFactory.getDictionary(dictFile, locale)` can still return null, e.g. with a bad dictionary file.
* Tthe dictionary facilitator is now static, so it should not use the `m` that indicates instance variables. Choose whether you use `s` or not, I just find the `m` misleading here.

(I'll go though whatever came up in the last 2 days and hope to merge this tomorrow at latest)

Helium314 commented:
> Before releasing this, the dictionaries have to be made available and added to dictionaries_in_dict_repo.csv, right?

Definitely, and the dictionary screen needs to be adapted so users can disable the emoji dictionary for normal suggestions (or maybe set the weight, with 0 just disabling the dictionary).
But for that there is enough time between 3.2 and 3.3 beta.

eranl commented:
Are you planning all of that for a separate PR? Which part can I start working on?

Helium314 commented:
`dictionaries_in_dict_repo.csv` is updated right before a release.
So what needs to be done before the first 3.3 beta is
* merge this PR
* update the dictionary screen to expose the weight (there is some todo in `DictionaryFactory` I think)
* finish the dictionary PR
* and I think the inline emoji search should also be added in 3.3

So all you can start working on is the dictionary screen update, as you're already working on the rest.

Helium314 commented:
> Dictionary loading will not work for all languages. Should I add [this change](https://github.com/Helium314/HeliBoard/blob/ab927aea2df7bbc1bb56ba305c40ca74d793bce7/app/src/main/java/helium314/keyboard/latin/utils/DictionaryInfoUtils.kt#L108-L112) in a separate PR?

I don't think locales should be mixed that way. If a dictionary gets added to an incorrect directory, then I think it's better to fix adding dictionaries than trying to deal with this issue.

> Should this be done in a separate PR or in the emoji search one?

In a separate PR, because it's not only for emoji dictionaries but for any dictionary.

> How about adding a global boolean _Use emoji dictionaries for regular suggestions_ setting?

You mean instead of setting a weight, or additional? With additional allowing users to disable emoji suggestions in 2 different ways can be confusing. Though we could make the globar switch change the weight, so things should be clear enough.

Helium314 commented:
> What do you mean by mixed? Isn't this normal locale fallback logic?

Dictionaries should always be in the cached folder of the locale they were added to. This is how it already works when extracting assets dictionaries, and I think it should stay that way. The fallback logic is a bit more more exhaustive (using the _bestMatch_) and decides which assets dictionary is extracted into the cache folder for the current locale.

> What's the fix you're referring to?

I don't know the bug, so I can't tell you. But you imply that dictionaries are added to the wrong locale for some reason, and that should not happen.

> I meant instead, as a simplification. If you prefer the general weight approach, I'd be happy to implement, but would need some guidance about the details you're planning.

Weight has the advantage that people can adjust how many emojis they get with the normal dictionary, instead of being stuck with f <= 20. Might also be useful for some other addon dictionaries.
But there is certainly no _need_ to add it, we can also start with the simple setting.

eranl commented:
> I don't know the bug, so I can't tell you. But you imply that dictionaries are added to the wrong locale for some reason, and that should not happen.

As I described [here](https://github.com/Helium314/HeliBoard/pull/1660#issuecomment-2952517726), the *Lao (Laos)* language's locale in HeliBoard is `lo_LA`, and the CLDR-based dictionary is for `lo`.

eranl commented:
The main issue with dictionary reloading is that it relies on `isInitialized`, which is now not a sufficient condition - we now have to check if the group contains an emoji dict, which is hidden by the `mainDict` collection.

eranl commented:
> > we can also start with the simple setting.

> > Either way, the logic for determining if the dictionary group has to be reloaded, and managing the reloading, becomes more complex. Do you have any tips for that?
> 
> I'm not sure what exactky you are referring to? _When_ to reload the collection?

> The main issue with dictionary reloading is that it relies on `isInitialized`, which is now not a sufficient condition - we now have to check if the group contains an emoji dict, which is hidden by the `mainDict` collection.

This is a summary of our discussion. When the user changes the new setting, we have to reload the dictionary group. In order to detect that, we have to check if the group contains an emoji dict.

Helium314 commented:
We can just reload the dictionaries anyway, it's not like the user will change settings every few minutes.
